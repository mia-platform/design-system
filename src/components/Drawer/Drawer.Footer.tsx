/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement } from 'react'
import { isEmpty } from 'lodash-es'

import { CustomDrawerFooter, DrawerFooter, FooterProps } from './Drawer.types'
import styles from './Drawer.module.css'

function isDrawerFooter(obj: DrawerFooter | CustomDrawerFooter): obj is DrawerFooter {
  return (
    obj
    && ('buttons' in obj || 'extra' in obj)
  )
}

export const Footer = ({ footer }: FooterProps): ReactElement | null => {
  if (!footer || isEmpty(footer)) {
    return null
  }

  if (isDrawerFooter(footer)) {
    const { buttons, extra } = footer
    return <footer className={styles.footer}>
      {(buttons || extra) && <>
        <div className={styles.extra}>{extra}</div>
        <div className={styles.footerButtons}>
          {buttons}
        </div>
      </>}
    </footer>
  }

  return footer
}

