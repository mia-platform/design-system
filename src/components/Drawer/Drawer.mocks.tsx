/* istanbul ignore file */
/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

// This is a file containing mock components and helpers
/* eslint-disable react/no-multi-comp */

import { ReactElement } from 'react'

import { Button } from '../Button'
import { Hierarchy } from '../Button/Button.types'

export const DrawerLipsum = (): ReactElement => {
  return (
    <div style={{ display: 'flex', flexDirection: 'column', gap: 24 }}>
      <span>
        {`Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla a semper diam, nec congue lectus. Mauris pellentesque convallis augue eu ornare. Sed pulvinar velit ut congue vulputate. Etiam elementum lacus vitae felis vulputate ultricies. Vivamus eu erat orci. Proin a accumsan nunc, eget rutrum arcu. Donec efficitur justo vel porta placerat. Nunc euismod at nisi vel interdum. Morbi id pellentesque quam. Donec eleifend nisi feugiat est rutrum dapibus quis at urna. Ut vehicula hendrerit neque, vel ultrices diam dignissim id. Aliquam erat volutpat. Sed porta iaculis dolor, ac egestas mi accumsan quis. Nam justo ex, facilisis in nunc id, suscipit aliquam arcu.`}
      </span>
      <span>
        {'Nullam tempor nisi nulla, ac lacinia dolor porta in. Nam non velit enim. Nullam a orci id diam pretium sollicitudin. Nunc vitae mauris pretium, finibus orci non, varius diam. Donec sollicitudin, libero non blandit pellentesque, dolor neque elementum ante, in molestie massa ipsum sed odio. Morbi vestibulum pretium tempor. Suspendisse placerat faucibus velit, ac condimentum eros placerat sit amet. Etiam non finibus ipsum, vitae sagittis lorem. Curabitur convallis nisl sapien, quis placerat metus tempus ac. Maecenas molestie neque lorem, non ornare risus egestas ut.'}
      </span>
    </div>
  )
}

export const DrawerLipsumTitle = (): ReactElement => {
  return <span>{'Drawer Lipsum'}</span>
}

export const DrawerLipsumFooterButton = ({ text, hierarchy } : {text: string, hierarchy?: Hierarchy}): ReactElement => {
  return (
    <Button hierarchy={hierarchy}>
      {text}
    </Button>
  )
}

export const drawerLipsumFooter = {
  buttons: [
    <DrawerLipsumFooterButton key="action-button" text="Primary Action" />,
    <DrawerLipsumFooterButton hierarchy={Hierarchy.Neutral} key="action-button" text="Secondary Action" />,
  ],
  extra: 'Extra text',
}
