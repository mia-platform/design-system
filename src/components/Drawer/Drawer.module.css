.drawer {

  :global(.mia-platform-drawer-header-title) {
    flex-direction: row-reverse;

    :global(.mia-platform-drawer-close) {
      margin: 0;
      padding: 0;
    }

    h4 {
      margin: 0;
    }
  }

  :global(.mia-platform-drawer-footer) {
    padding: var(--spacing-padding-lg, 16px) var(--spacing-padding-xl, 24px);
    min-height: var(--shape-size-xl, 32px);
  }

  .footer {
    display: flex;
    align-items: center;
    justify-content: end;
    gap: var(--spacing-gap-sm, 8px);
  }

  .footerButtons {
    margin-inline-start: 0;
    display: flex;
    flex-direction: row-reverse;
    gap: var(--spacing-gap-sm, 8px);

    :global(*) {
      margin: 0 !important;
    }
  }

  .extra {
    flex: 1;
  }
}
