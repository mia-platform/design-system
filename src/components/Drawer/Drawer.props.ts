/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

export type DrawerProps = {

    /**
     * The children nodes to be rendered within the Drawer body.
     */
    children?: ReactNode,

    /**
     * Whether to unmount child components on close.
     */
    destroyOnClose?: boolean,

    ref?: React.ForwardedRef<HTMLDivElement>,

    /**
     * Drawer footer.
     */
    footer?: ReactNode,

    id?: string,

    key?: string,

    /**
     * Whether the modal is visible.
     */
    isVisible?: boolean,

    /**
     * Function invoked at the click of the Drwaer close (X) button.
     */
    onClose?: () => void,

    /**
     * Title of the drawer, which briefly conveys information about its contents.
     */
    title: ReactNode,
}
