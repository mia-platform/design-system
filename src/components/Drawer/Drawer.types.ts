import { ReactElement, ReactNode } from 'react'

// Header

export type DrawerTitle = ReactNode

export type TitleProps = {
  docLink?: string,
  title: DrawerTitle,
}

// Footer

export type DrawerFooter = {

  /**
  * Array of buttons to be displayed in the right side of the footer.
  * Note that the rendering order is the opposite of the list order.
   */
  buttons?: ReactElement[]

  /**
  * Extra information to be displayed in the left side of the footer (such as text or a checkbox).
   */
  extra?: ReactNode
}

export type CustomDrawerFooter = ReactElement

export type FooterProps = {
    footer?: DrawerFooter | CustomDrawerFooter,
}
