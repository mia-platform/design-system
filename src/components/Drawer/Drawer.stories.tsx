/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Meta, StoryObj } from '@storybook/react'
import { useArgs } from '@storybook/preview-api'

import { DrawerLipsumFooterButton, DrawerLipsumTitle, drawerLipsumFooter } from './Drawer.mocks'
import { Button } from '../Button'
import { Drawer } from '.'

const defaults = {
  title: <DrawerLipsumTitle />,
}

const meta = {
  component: Drawer,
  args: defaults,
  argTypes: {
    children: { control: false },
  },
  decorators: [
    (Story) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const [_, setArgs] = useArgs()
      return <div>
        <Button
          onClick={() => setArgs(
            {
              isVisible: true,
              onClose: () => { setArgs({ isVisible: false }) },
            }
          )}
        >
          Open drawer
        </Button>
        <Story />
      </div>
    },
  ],
} satisfies Meta<typeof Drawer>

export default meta
type Story = StoryObj<typeof meta>

export const BasicExample: Story = {}

export const WithStandardFooter: Story = {
  args: {
    ...meta.args,
    footer: drawerLipsumFooter,
  },
}

export const WithCustomFooter: Story = {
  args: {
    ...meta.args,
    footer: <DrawerLipsumFooterButton text="Custom footer button" />,
  },
}
