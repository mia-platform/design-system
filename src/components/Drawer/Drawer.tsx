/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Drawer as AntdDrawer } from 'antd'
import { ReactElement } from 'react'

import { DrawerProps } from './Drawer.props'
import { Footer } from './Drawer.Footer'
import { Title } from './Drawer.Title'

const styles = {
  footer: { padding: '24px' },
}

const DRAWER_WIDTH = 512

export const Drawer = ({
  children,
  destroyOnClose,
  footer,
  id,
  isVisible,
  key,
  onClose,
  title,
}: DrawerProps): ReactElement => {
  return (
    <AntdDrawer
      closeIcon={null}
      destroyOnClose={destroyOnClose}
      footer={<Drawer.Footer footer={footer} />}
      id={id}
      key={key}
      open={isVisible}
      styles={styles}
      title={<Drawer.Title title={title} />}
      width={DRAWER_WIDTH}
      onClose={onClose}
    >
      {children}
    </AntdDrawer>
  )
}

Drawer.Title = Title
Drawer.Footer = Footer
