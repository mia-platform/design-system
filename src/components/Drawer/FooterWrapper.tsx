import { ReactElement, ReactNode } from 'react'

import { DrawerProps } from './Drawer.props'

// export const FooterWrapper = ({ children }: {children: ReactNode}): ReactElement => {
//   return (
//     <div styles={}>
//       {children}
//     </div>
//   )
// }
