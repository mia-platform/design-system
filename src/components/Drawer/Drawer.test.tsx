/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { DrawerLipsum, DrawerLipsumFooterButton, DrawerLipsumTitle } from './Drawer.mocks'
import { render, screen } from '../../test-utils'
import { Drawer } from './Drawer'
import { DrawerProps } from './Drawer.props'

describe('Drawer', () => {
  const props: DrawerProps = {
    children: 'Drawer Content',
    footer: <DrawerLipsumFooterButton />,
    isVisible: true,
    title: <DrawerLipsumTitle />,
    onClose: jest.fn(),
  }

  beforeEach(() => jest.resetAllMocks())

  it('renders drawer with provided title and footer', () => {
    const { baseElement } = render(<Drawer {...props} ><DrawerLipsum /></Drawer>)

    expect(screen.getByText(/drawer lipsum/i)).toBeVisible()
    expect(screen.getByRole('button', { name: /close/i })).toBeInTheDocument()
    expect(screen.getByText(/Lorem ipsum dolor sit amet,/i)).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
  })

  it('does not render drawer when isVisible is false', () => {
    render(<Drawer {...props} isVisible={false} >{'the-content'}</Drawer>)
    expect(screen.queryByText(/drawer lipsum/i)).toBeNull()
  })
})
