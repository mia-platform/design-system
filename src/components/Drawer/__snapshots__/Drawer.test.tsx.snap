// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer renders drawer with provided title and footer 1`] = `
<body>
  <div />
  <div>
    <div
      class="mia-platform-drawer mia-platform-drawer-right mia-platform-drawer-open"
      tabindex="-1"
    >
      <div
        class="mia-platform-drawer-mask"
      />
      <div
        aria-hidden="true"
        data-sentinel="start"
        style="width: 0px; height: 0px; overflow: hidden; outline: none; position: absolute;"
        tabindex="0"
      />
      <div
        class="mia-platform-drawer-content-wrapper"
        style="width: 512px;"
      >
        <div
          aria-modal="true"
          class="mia-platform-drawer-content"
          role="dialog"
        >
          <div
            class="mia-platform-drawer-header"
          >
            <div
              class="mia-platform-drawer-header-title"
            >
              <div
                class="mia-platform-drawer-title"
              >
                <h4
                  aria-label=""
                  class="mia-platform-typography mia-platform-typography-ellipsis mia-platform-typography-single-line mia-platform-typography-ellipsis-single-line"
                  role="h4"
                >
                  <span>
                    Drawer Lipsum
                  </span>
                </h4>
              </div>
            </div>
          </div>
          <div
            class="mia-platform-drawer-body"
          >
            <div
              style="display: flex; flex-direction: column; gap: 24px;"
            >
              <span>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla a semper diam, nec congue lectus. Mauris pellentesque convallis augue eu ornare. Sed pulvinar velit ut congue vulputate. Etiam elementum lacus vitae felis vulputate ultricies. Vivamus eu erat orci. Proin a accumsan nunc, eget rutrum arcu. Donec efficitur justo vel porta placerat. Nunc euismod at nisi vel interdum. Morbi id pellentesque quam. Donec eleifend nisi feugiat est rutrum dapibus quis at urna. Ut vehicula hendrerit neque, vel ultrices diam dignissim id. Aliquam erat volutpat. Sed porta iaculis dolor, ac egestas mi accumsan quis. Nam justo ex, facilisis in nunc id, suscipit aliquam arcu.
              </span>
              <span>
                Nullam tempor nisi nulla, ac lacinia dolor porta in. Nam non velit enim. Nullam a orci id diam pretium sollicitudin. Nunc vitae mauris pretium, finibus orci non, varius diam. Donec sollicitudin, libero non blandit pellentesque, dolor neque elementum ante, in molestie massa ipsum sed odio. Morbi vestibulum pretium tempor. Suspendisse placerat faucibus velit, ac condimentum eros placerat sit amet. Etiam non finibus ipsum, vitae sagittis lorem. Curabitur convallis nisl sapien, quis placerat metus tempus ac. Maecenas molestie neque lorem, non ornare risus egestas ut.
              </span>
            </div>
          </div>
          <div
            class="mia-platform-drawer-footer"
            style="padding: 24px;"
          >
            <footer>
              <div>
                <button
                  class="mia-platform-btn mia-platform-btn-primary button buttonMd"
                  type="button"
                >
                  <div
                    class="buttonText"
                  >
                    Close
                  </div>
                </button>
              </div>
            </footer>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-sentinel="end"
        style="width: 0px; height: 0px; overflow: hidden; outline: none; position: absolute;"
        tabindex="0"
      />
    </div>
  </div>
</body>
`;
